dbms/src/Debug/MockRaftStoreProxy.cpp

```diff
@@ -295,6 +326,70 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
     cb(*this);
 }
 
+void MockRaftStoreProxy::bootstrap(
+    const Context & ctx,
+    UInt64 region_id)
+{
+    auto _ = genLockGuard();
+    regions.emplace(region_id, std::make_shared<MockProxyRegion>(region_id));
+
+    KVStore & kvs = *ctx.getTMTContext().getKVStore();
+    auto task_lock = kvs.genTaskLock();
+    auto lock = kvs.genRegionWriteLock(task_lock);
+    {
+        auto region = tests::makeRegion(region_id, RecordKVFormat::genKey(region_id, 0), RecordKVFormat::genKey(region_id, 10));
+        lock.regions.emplace(region_id, region);
+        lock.index.add(region);
+    }
+}
+
+void MockRaftStoreProxy::normalWrite(
+    const Context & ctx,
+    const FailCond & cond,
+    UInt64 region_id,
+    std::vector<HandleID> keys,
+    std::vector<std::string> vals,
+    std::vector<WriteCmdType> cmd_types,
+    std::vector<ColumnFamilyType> cmd_cf)
+{
+    auto region = getRegion(region_id);
+    assert(region != nullptr);
+    auto index = region->getLatestCommitIndex() + 1;
+    auto term = region->getLatestCommitTerm();
+    raft_cmdpb::RaftCmdRequest request;
+    size_t n = keys.size();
+    assert(n == vals.size());
+    assert(n == cmd_types.size());
+    assert(n == cmd_cf.size());
+    for (size_t i = 0; i < n; i++)
+    {
+        if (cmd_types[i] == WriteCmdType::Put)
+        {
+            auto cf_name = CFToName(cmd_cf[i]);
+            auto key = RecordKVFormat::genKey(1, keys[i], 1);
+            TiKVValue value = std::move(vals[i]);
+            RegionBench::setupPutRequest(request.add_requests(), cf_name, key, value);
+        }
+        else
+        {
+            auto cf_name = CFToName(cmd_cf[i]);
+            auto key = RecordKVFormat::genKey(1, keys[i], 1);
+            RegionBench::setupDelRequest(request.add_requests(), cf_name, key);
+        }
+    }
+
+    if (cond.fail_before_kvstore) return;
```

Comment:
maybe use FailPointHelper and fiu_do_on instead of passing a `FailCond` to mock the failure?

Reply:
I am afraid that failpoints take its effects on the whole program, by giving a explicit fail condition for every write can provide more flexibility in such tests.